2-structured-data/test/crud.test.js

```diff
@@ -0,0 +1,196 @@
+      var expected = '<div class="media-body"><h4>my book</h4><p></p></div>';
+      request(proxyquire('../app', stubs))
+        .get('/books')
+        .expect(200)
+        .expect(function (response) {
+          assert.ok(response.text.indexOf(expected) !== -1);
+        })
+        .end(done);
+    });
+
+    it('should handle error', function (done) {
+      var expected = 'Bad Request';
+      if (process.version.indexOf('v0.10') !== -1) {
+        expected = 'Error during request.';
+      }
+      if (require('../config')().dataBackend === 'cloudsql') {
+        expected = 'ER_SP_UNDECLARED_VAR: Undeclared variable: NaN';
+      }
+      request(proxyquire('../app', stubs))
+        .get('/books')
+        .query({ pageToken: 'badrequest' })
+        .expect(500)
+        .expect(function (response) {
+          assert.ok(response.text.indexOf(expected) !== -1);
+        })
+        .end(done);
+    });
+
+    // delete the book
+    after(function (done) {
+      if (id) {
+        request(proxyquire('../app', stubs))
+          .delete('/api/books/' + id)
+          .expect(200)
+          .end(done);
+      } else {
+        done();
+      }
+    });
+  });
+
+  describe('/books/add', function () {
+    var id;
+
+    it('should post to add book form', function (done) {
+      request(proxyquire('../app', stubs))
+        .post('/books/add')
+        .send('title=my%20book')
+        .expect(302)
+        .expect(function (response) {
+          var location = response.headers.location;
+          id = parseInt(location.replace('/books/', ''), 10);
+          assert.ok(response.text.indexOf('Redirecting to /books/') !== -1);
+        })
+        .end(done);
+    });
+
+    it('should show add book form', function (done) {
+      request(proxyquire('../app', stubs))
+        .get('/books/add')
+        .expect(200)
+        .expect(function (response) {
+          assert.ok(response.text.indexOf('Add book') !== -1);
+        })
+        .end(done);
+    });
+
+    // delete the book
+    after(function (done) {
+      if (id) {
+        request(proxyquire('../app', stubs))
+          .delete('/api/books/' + id)
+          .expect(200)
+          .end(done);
+      } else {
+        done();
+      }
+    });
+  });
+
+  describe('/books/:book/edit & /books/:book', function () {
+    var id;
+
+    // setup a book
+    before(function (done) {
+      request(proxyquire('../app', stubs))
+        .post('/api/books')
+        .send({ title: 'my book' })
+        .expect(200)
+        .expect(function (response) {
+          id = response.body.id;
+          assert.ok(response.body.id);
+          assert.equal(response.body.title, 'my book');
+        })
+        .end(done);
+    });
+
+    it('should update a book', function (done) {
+      var expected = 'Redirecting to /books/' + id;
+      request(proxyquire('../app', stubs))
+        .post('/books/' + id + '/edit')
+        .send('title=my%20other%20book')
```

Comment:
I believe the test is slightly different in step 2 because step 2 itself uses different form parser middleware than the rest of the steps. I'm not sure why.